### PR TITLE
Bootstrap associative cf arithmetic

### PR DESCRIPTION
**Pull request by kostrzewa**
This is a work-in-progress PR to consistently operate on the bootstrap samples under associative modifications of a correlation function.

### CODE REVIEW

**Comment by kostrzewa** [comment]
Note that this contains commits from the `icf_support` branch to avoid duplication of effort. That branch should be merged first via #186

**Comment by kostrzewa** [comment]
There's an annoying issue that we have to take care of somehow. When `tsboot` is used, `cf$cf.tsboot` and `cf$icf.tsboot` contain the original data in their `tseries` member.

On the other hand, when `jackknife.cf` is used, the "hacked-together" `cf.tsboot` does not contain the original data.

In order to work consistently on the samples, I need to address this and see two alternatives:

1) we don't propagate the full result of `tsboot` but only a subset of its members
2) we add the `tseries` member to what `jackknife.cf` attaches

**Comment by kostrzewa** [comment]
I also think that the value of `sim` under `jackknife.cf` should be `fixed`, but that's only cosmetic anyway.

**Comment by kostrzewa** [comment]
Only `symmetrise.cf` left to go.

**Comment by kostrzewa** [comment]
Merge depends on #186

**Comment by kostrzewa** [comment]
ready for review

**Comment by kostrzewa** [comment]
ugh, because I've removed a function, the diff for `R/cf.R` is basically useless to see what has changed... first time I see this here

**Comment by kostrzewa** [comment]
almost a month has passed, requesting a review again

File: R/cf.R [hunk 1]
```diff
@@ -623,15 +671,52 @@ add.cf <- function(cf1, cf2, a = 1.0, b = 1.0) {
...
 
   if( has_icf(cf1) | has_icf(cf2) ){
     stopifnot( has_icf(cf1) & has_icf(cf2) )
+    stopifnot( all(dim(cf1$icf) == dim(cf2$icf) ) )
     cf$icf <- a*cf1$icf + b*cf2$icf
   }
 
-  cf <- invalidate.samples.cf(cf)
+  # now reconstruct the bootstrap samples
+  invalidate.samples.cf(cf)
```
Comment (martin-ueding) [inline R/cf.R]: This statement does not have effect, right?

Comment (kostrzewa) [reply R/cf.R]: The idea, as elsewhere in the code, was to ensure a completely clean slate before re-adding the updated bootstrap samples via the constructor.

Comment (martin-ueding) [reply R/cf.R]: I realize that part. But without the `cf <-` assignment that line is without an effect due to the pass-b-promise nature of R.

Comment (kostrzewa) [reply R/cf.R]: Now I see what you mean, my brain didn't even parse that the assignment had been removed...

**Pull request merged**